Add derivatives to material outputs

MFraters opened the pull request:
Implemented only the bare minimum for the stokes equation. 

Viscosity is dependent on velocity, pressure and strain rate. 
Strain rate is dependent on velocity.
Density is dependent on pressure.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -245,6 +245,39 @@ namespace aspect
+	/**
+	 * The derivatives of the strain rates
+	 */
+    	std::vector<SymmetricTensor<2,dim> > dstrain_rate_du;
```
You won't need this variable. In fact, you can't compute it because the strain rate is not a function of the velocity but of the derivatives of the velocity instead. But I'll show you tomorrow why you won't need it either.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -365,6 +398,11 @@ namespace aspect
+      /**
+       * The derivatives used for forming the Jacobian, needed for the Newton solver.
+       */
+      MaterialModelDerivatives<dim> derivatives;
```
@tjhei and @jdannberg : Do you think this is the right design to have a member variable in the `MaterialModelOutputs` variable, or would you prefer a `std_cxx11::shared_ptr<...>` pointer to a derivative object?

I figure that you probably want to have something similar in there for melt transport, and maybe it would be easier to have a collection of pointers to objects derived from a common base class, rather than storing a whole bunch of member variables here, all of which will rarely even be used.

jdannberg replied:
We were thinking about a collection of shared pointers to objects derived from a common base class (or at least that was Timo's idea) and we already started to work on something like that. So we should probably all sit together and talk at some point.

bangerth replied:
On 01/13/2016 02:53 PM, Juliane Dannberg wrote:
> We were thinking about a collection of shared pointers to objects
> derived from a common base class (or at least that was Timo's idea) and
> we already started to work on something like that. So we should probably
> all sit together and talk at some point.
Come down here at 4 :-)

bangerth commented on source/material_model/interface.cc:
```diff
@@ -281,8 +281,24 @@ namespace aspect
+    MaterialModelDerivatives<dim>::MaterialModelDerivatives(const unsigned int n_points,
+            const unsigned int n_comp)
+    {
+       dviscosities_du.resize(n_points, aspect::Utilities::signaling_nan<double>());
```
Please move this into a member initializer list (i.e., what comes after the `:` in a constructor call). I know this is the same style as is currently used below in `MaterialModelOutputs`, where we should probably change it as well.

bangerth commented on source/material_model/interface.cc:
```diff
@@ -281,8 +281,24 @@ namespace aspect
     MaterialModelOutputs<dim>::MaterialModelOutputs(const unsigned int n_points,
-                                                    const unsigned int n_comp)
+                                                    const unsigned int n_comp) 
+    : derivatives(false?n_points:0, false?n_comp:0)
```
This isn't particularly useful right now :-)

MFraters replied:
true, I forgot that I had to work on that a bit more..

MFraters commented:
Added the diffusion_dislocation_plasticity material model with the analytical derivatives. Also added a test for the simpler material model derivatives. @bangerth Is this the structure in the test you where talking about?

The test itself doesn't work yet because the linker apparently can't find the MaterialModelDerivatives in line 59: out_basis.additional_outputs.push_back(std::make_shared<MaterialModelDerivatives<dim> > (3, 3));, but it seems to be able to find it three line before.

bangerth commented on include/aspect/material_model/diffusion_dislocation_plasticity.h:
```diff
@@ -0,0 +1,192 @@
+         * or as $\nabla \cdot \mathbf{u}=0$ (incompressible Stokes).
+        *
+        * This material model is incompressible.
+         */
```
Please align the asterisks.

bangerth commented on include/aspect/material_model/diffusion_dislocation_plasticity.h:
```diff
@@ -0,0 +1,192 @@
+
+        /*double average_value (const std::vector<double> &composition,
+                              const std::vector<double> &parameter_values,
+                              const double &composition_averaging_exponent) const;*/
```
If you don't need the function, just remove it.

bangerth commented on include/aspect/material_model/diffusion_dislocation_plasticity.h:
```diff
@@ -0,0 +1,192 @@
+                                                                  std::vector<double> &stress_pressure_deriv) const;
+
+
+        std::vector<double> yield_stress;
```
For this vector and those below, are they all per-composition? If so, maybe add a comment before the first of these variables to explain this.

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -398,10 +432,19 @@ namespace aspect
-       * Constant version of get_additional_output() returning a const pointer.
-       */
-      template <class AdditionalOutputType>
-      const AdditionalOutputType *get_additional_output() const;
```
Any particular reason to remove this function?

bangerth commented on include/aspect/material_model/interface.h:
```diff
@@ -398,10 +432,19 @@ namespace aspect
+              */
+      void create_additional_material_outputs(const unsigned int n_points,
+                                              const unsigned int n_comp)
+      {
```
I think we probably need to move this somewhere else. But this can wait.

bangerth commented on source/material_model/diffusion_dislocation_plasticity.cc:
```diff
@@ -0,0 +1,665 @@
+/*
+  Copyright (C) 2011 - 2015 by the authors of the ASPECT code.
```
Fix copyright year.

bangerth commented on source/material_model/diffusion_dislocation_plasticity.cc:
```diff
@@ -0,0 +1,665 @@
+    /* template <int dim>
+     double
+     DiffusionDislocationPlasticity<dim>::
+     average_value ( const std::vector<double> &composition,
```
If you don't need the function, just remove it.

bangerth commented on source/material_model/simpler.cc:
```diff
@@ -65,6 +73,15 @@ namespace aspect
           out.thermal_conductivities[i] = k_value;
           out.compressibilities[i] = 0.0;
+
+          derivatives->dviscosities_du[i] = 0.0;
```
You can only run this block of lines if `derivatives != NULL`.

MFraters replied:
I am not sure what you mean, because the derivative is zero in this case. Should I just not set it and leave it at the default value of signaling:nan?

bangerth replied:
What is meant is that the function also needs to successfully run if the caller is not actually interested in derivatives. In that case, `derivatives` is just a NULL pointer and you should not dereference it. So you need a
```
  // see if we need to compute derivatives
  if (derivatives != 0)
   {
     ...
   }
```
block around lines 77 to 84.

MFraters replied:
ahh, I get it!

bangerth commented on tests/material_derivatives_simpler.cc:
```diff
@@ -0,0 +1,95 @@
+
+  in_basis.strain_rate[0] = SymmetricTensor<2,dim>(); // What do I fill it with?
+  in_basis.strain_rate[1] = SymmetricTensor<2,dim>(); // What do I fill it with?
+  in_basis.strain_rate[2] = SymmetricTensor<2,dim>(); // What do I fill it with?
```
I think this is reasonable. Or, and maybe better, just choose a nonzero tensor with entries that are about the order of magnitude we would expect in the earth.

bangerth commented on tests/material_derivatives_simpler.cc:
```diff
@@ -0,0 +1,95 @@
+  in_dstrainrate = in_basis;
+  in_dstrainrate.strain_rate[0] *=1.0000000001;
+  in_dstrainrate.strain_rate[1] *=1.0000000001;
+  in_dstrainrate.strain_rate[2] *=1.0000000001;
```
This multiplies *all* components of the strain rate tensor by the same factor. But you probably want to test derivatives with regard to the components of the strain rate *separately*, so you need `dim*(dim+1)/2` different `in_dstrainrate` objects, and in each of them you multiply only one component by 1.0000001.

bangerth commented on tests/material_derivatives_simpler.cc:
```diff
@@ -0,0 +1,95 @@
+
+  MaterialModelOutputs<dim> out_basis(3,3);
+  MaterialModelOutputs<dim> out_dp(3,3);
+  MaterialModelOutputs<dim> out_dstrainrate(3,3);
```
And correspondingly you need multiple such output objects here.

bangerth commented on tests/material_derivatives_simpler.cc:
```diff
@@ -0,0 +1,95 @@
+  const int dim=2;
+std::cout << "Flag 1" << std::endl;
+  using namespace aspect::MaterialModel;
+  MaterialModelInputs<dim> in_basis(3,3);
```
The proper term is the "base case", so maybe `in_base` instead?

MFraters commented:
I got the material_derivatives_simpler test working now on the linux cluster, but on my mac I get the error that it can't find the MaterialModelDerivatives symbol. Any idea's where this error is coming from?

```
test 134
    Start 134: material_derivatives_simpler

134: Test command: /opt/local/bin/cmake "-DBINARY_DIR=/Users/Menno/Documents/Phd/aspect/aspect_2015-05-22/aspect/build" "-DTESTNAME=tests.material_derivatives_simpler" "-DERROR="Test material_derivatives_simpler failed"" "-P" "/Users/Menno/Documents/Phd/aspect/aspect_2015-05-22/aspect/tests/run_test.cmake"
134: Test timeout computed to be: 600
134: dyld: lazy symbol binding failed: Symbol not found: __ZN6aspect13MaterialModel24MaterialModelDerivativesILi2EEC1Ejj
134:   Referenced from: ./libmaterial_derivatives_simpler.so
134:   Expected in: flat namespace
134: 
134: dyld: Symbol not found: __ZN6aspect13MaterialModel24MaterialModelDerivativesILi2EEC1Ejj
134:   Referenced from: ./libmaterial_derivatives_simpler.so
134:   Expected in: flat namespace
134: 
134: [0]PETSC ERROR: ------------------------------------------------------------------------
134: [0]PETSC ERROR: Caught signal number 5 TRAP
134: [0]PETSC ERROR: Try option -start_in_debugger or -on_error_attach_debugger
134: [0]PETSC ERROR: or see http://www.mcs.anl.gov/petsc/documentation/faq.html#valgrind
134: [0]PETSC ERROR: or try http://valgrind.org on GNU/linux and Apple Mac OS X to find memory corruption errors
134: [0]PETSC ERROR: configure using --with-debugging=yes, recompile, link, and run 
134: [0]PETSC ERROR: to get more information on the crash.
134: [0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
134: [0]PETSC ERROR: Signal received
134: [0]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
134: [0]PETSC ERROR: Petsc Release Version 3.5.4, unknown 
134: [0]PETSC ERROR: ../aspect on a uly named Mennos-MacBook-Pro.local by Menno Mon Jan 25 18:57:46 2016
134: [0]PETSC ERROR: Configure options --with-make-np=8 --with-debugging=0 --prefix=/Applications/deal.II.app/Contents/Resources/opt/petsc-9fbf19a --with-mpi-dir=/Applications/deal.II.app/Contents/Resources/opt/openmpi-1.6.5 --with-sundials-dir=/Applications/deal.II.app/Contents/Resources/opt/sundials-2.6.1 --with-shared-libraries --with-external-packages-dir=/Users/heltai/cluster-config/petsc-external --download-parmetis --download-metis --download-hypre --download-mumps --download-scalapack --download-superlu --download-superlu_dist --download-hdf5
134: [0]PETSC ERROR: #1 User provided function() line 0 in  unknown file
134: --------------------------------------------------------------------------
134: MPI_ABORT was invoked on rank 0 in communicator MPI_COMM_WORLD 
134: with errorcode 59.
134: 
134: NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
134: You may or may not see output from other processes, depending on
134: exactly when Open MPI kills them.
134: --------------------------------------------------------------------------
134: make[3]: *** [tests/output-material_derivatives_simpler/screen-output] Error 59
134: make[2]: *** [tests/CMakeFiles/material_derivatives_simpler.screen-output.diff.dir/all] Error 2
134: make[1]: *** [tests/CMakeFiles/tests.material_derivatives_simpler.dir/rule] Error 2
134: make: *** [tests.material_derivatives_simpler] Error 2
134: *** "Test material_derivatives_simpler failed": ***
134: Scanning dependencies of target material_derivatives_simpler
134: [  0%] Building CXX object tests/CMakeFiles/material_derivatives_simpler.dir/material_derivatives_simpler.cc.o
134: [  0%] Linking CXX shared library libmaterial_derivatives_simpler.so
134: [  0%] Built target material_derivatives_simpler
134: [100%] Built target aspect
134: [100%] Generating output-material_derivatives_simpler/screen-output
134: 
134: CMake Error at /Users/Menno/Documents/Phd/aspect/aspect_2015-05-22/aspect/tests/run_test.cmake:14 (MESSAGE):
134:   *** Test aborted.
134: 
134: 
1/1 Test #134: material_derivatives_simpler .....***Failed   16.96 sec
```

bangerth commented:
You need to explicitly instantiate the `MaterialOutputDerivatives` class using the statement
```
  template class MaterialOutputDerivatives<dim>;
```
somewhere at the bottom of the file where its functions are defined (and where there are other explicit instantiations already).

MFraters commented:
Thanks! It works now.

gassmoeller commented:
Superseded by #1683. If not, feel free to reopen.
